invenio_rdm_records/services/access/service.py

```diff
@@ -353,37 +359,66 @@ def _check_grant_subject(self, identity, grant):
             return False
 
     @unit_of_work()
-    def create_grant(self, identity, id_, data, expand=False, uow=None):
-        """Create an access grant for a record (resp. its parent)."""
+    def bulk_create_grants(self, identity, id_, data, expand=False, uow=None):
+        """Bulk create access grants for a record (resp. its parent)."""
         record, parent = self.get_parent_and_record_or_draft(id_)
 
         # Permissions
         self.require_permission(identity, "manage", record=record)
 
         # Validation
-        data, __ = self.schema_grant.load(
+        data, __ = self.schema_grants.load(
             data, context={"identity": identity}, raise_errors=True
         )
 
-        # Creation
-        grant = parent.access.grants.create(
-            subject_type=data["subject"]["type"],
-            subject_id=data["subject"]["id"],
-            permission=data["permission"],
-            origin=data.get("origin"),
-        )
+        grants = data["grants"]
+
+        new_grants = []
+
+        # fail if any of the grants already exist
+        if any(
+            existing_grant.subject_id == grant["subject"]["id"]
+            and existing_grant.subject_type == grant["subject"]["type"]
+            for existing_grant in parent.access.grants
+            for grant in grants
+        ):
+            raise GrantExistsError()
+
+        for grant in grants:
+            # Creation
+            new_grant = parent.access.grants.create(
+                subject_type=grant["subject"]["type"],
+                subject_id=grant["subject"]["id"],
+                permission=grant["permission"],
+                origin=grant.get("origin"),
+            )
 
-        if not self._check_grant_subject(identity, grant):
-            raise ValidationError(
-                _("Could not find the specified subject."), field_name="subject.id"
+            if not self._validate_grant_subject(identity, new_grant):
```

Comment:
what happens if e.g. the second grant fails? everything is rolled back, also the first one and it is not created. Is the notification still sent?

Reply:
No, in this case the notification is also not sent

Reply:
@anikachurilova can we do the check before creating it?

Reply:
Discussed irl, leaving it as is